File: src/lib/dataStore.ts

import { AppState, Book, Customer, User } from "./types";

// Initial state with hardcoded admin user
export const initialState: AppState = {
  books: [],
  customers: [],
  currentUser: null,
  currentScreen: "login"
};

// Function to load data from localStorage (simulating file load)
export const loadData = (): AppState => {
  try {
    const savedBooks = localStorage.getItem('books');
    const savedCustomers = localStorage.getItem('customers');
    
    return {
      books: savedBooks ? JSON.parse(savedBooks) : [],
      customers: savedCustomers ? JSON.parse(savedCustomers) : [],
      currentUser: null,
      currentScreen: "login"
    };
  } catch (error) {
    console.error("Error loading data:", error);
    return initialState;
  }
};

// Function to save data to localStorage (simulating file save)
export const saveData = (books: Book[], customers: Customer[]): void => {
  try {
    localStorage.setItem('books', JSON.stringify(books));
    localStorage.setItem('customers', JSON.stringify(customers));
    console.log("Data saved successfully");
  } catch (error) {
    console.error("Error saving data:", error);
  }
};

// Function to authenticate a user
export const authenticateUser = (username: string, password: string, state: AppState): User | null => {
  // Check for admin credentials
  if (username === "admin" && password === "admin") {
    return { username, password, role: "owner" };
  }
  
  // Check for customer credentials
  const customer = state.customers.find(
    (c) => c.username === username && c.password === password
  );
  
  if (customer) {
    return { username, password, role: "customer" };
  }
  
  return null;
};

File: src/com/bookstore/util/DataStore.java

package com.bookstore.util;

import com.bookstore.model.Book;
import com.bookstore.model.Customer;
import com.bookstore.model.User;

import java.util.ArrayList;
import java.util.List;

/**
 * Manages data storage and retrieval for the bookstore application.
 */
public class DataStore {
    private List<Book> books;
    private List<Customer> customers;
    private User currentUser;
    private String currentScreen;

    // Singleton instance
    private static DataStore instance;

    /**
     * Private constructor for singleton pattern.
     */
    private DataStore() {
        this.books = new ArrayList<>();
        this.customers = new ArrayList<>();
        this.currentUser = null;
        this.currentScreen = "login";
    }

    /**
     * Gets the singleton instance of DataStore.
     */
    public static synchronized DataStore getInstance() {
        if (instance == null) {
            instance = new DataStore();
        }
        return instance;
    }

    /**
     * Loads data from files.
     */
    public void loadData() {
        FileIO fileIO = new FileIO();
        this.books = fileIO.loadBooks();
        this.customers = fileIO.loadCustomers();
    }

    /**
     * Saves data to files.
     */
    public void saveData() {
        FileIO fileIO = new FileIO();
        fileIO.saveBooks(this.books);
        fileIO.saveCustomers(this.customers);
    }

    /**
     * Authenticates a user.
     */
    public User authenticateUser(String username, String password) {
        // Check for admin credentials
        if (username.equals("admin") && password.equals("admin")) {
            return new User(username, password, User.UserRole.OWNER);
        }
        
        // Check for customer credentials
        for (Customer customer : customers) {
            if (customer.getUsername().equals(username) && 
                customer.getPassword().equals(password)) {
                return new User(username, password, User.UserRole.CUSTOMER);
            }
        }
        
        return null;
    }

    // Getters and setters
    public List<Book> getBooks() {
        return books;
    }

    public void setBooks(List<Book> books) {
        this.books = books;
    }

    public List<Customer> getCustomers() {
        return customers;
    }

    public void setCustomers(List<Customer> customers) {
        this.customers = customers;
    }

    public User getCurrentUser() {
        return currentUser;
    }

    public void setCurrentUser(User currentUser) {
        this.currentUser = currentUser;
    }

    public String getCurrentScreen() {
        return currentScreen;
    }

    public void setCurrentScreen(String currentScreen) {
        this.currentScreen = currentScreen;
    }
}
